[Merged by Bors] - feat(CategoryTheory/Limits/Shapes/Reflexive): the walking reflexive pair

robin-carlier opened the pull request:
- Define `WalkingReflexivePair`, the category indexing reflexive pair.
- Show that the forgetful functor from this category to `WalkingParallelPair` is `Final`.
- Show that coequalizers of reflexive pairs are equivalent to colimits of diagrams indexed by `WalkingReflexivePair`.
- Show that having coequalizers of all reflexive pairs is equivalent to having all `WalkingReflexivePair`-shaped colimits.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

- [x] depends on: #17723 

There are probably a few names that could be better, perhaps some statements would belong to a new `ReflexivePair` namespace?

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -173,6 +194,365 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+    (WalkingReflexivePairHom._sizeOf_inst X X).sizeOf (𝟙 X) = 1 + sizeOf X := by cases X <;> rfl
+
+/-- The forgetful functor forgetting the common section -/
+def forgetReflexion : WalkingParallelPair ⥤ WalkingReflexivePair :=
```
A better name should be found because this is not a forgetful functor. Depending on the chosen name, this may have to be put in the `WalkingReflexivePair` or `WalkingParallelPair` namespace.

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -173,6 +194,365 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+
+@[simp]
+theorem reflexivePair_reflexion_comp_left (F : WalkingReflexivePair ⥤ C) :
+    F.map reflexion ≫  F.map left = 𝟙 F.obj zero :=
```
```suggestion
    F.map reflexion ≫ F.map left = 𝟙 F.obj zero :=
```
There are many other similar spacing issues.

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -173,6 +194,365 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+    | right_reflexion => g ≫  (commonSection f g)
+    | left_reflexion => f ≫  (commonSection f g)
+  map_comp := by
+    rintro _ _ _ ⟨⟩ g <;> cases g <;> try {rfl} <;>
+    try {dsimp; simp [↑ section_comp_left, ↑ section_comp_right]; rfl} <;> try {simp}
```
```suggestion
    rintro _ _ _ ⟨⟩ g <;> cases g <;> aesop_cat
```
I do not know what it is purpose of these ` ↑`?

kmill replied:
`↑` causes `simp` to apply the lemma after simplifying a subterm (I believe this is the default, so it shouldn't be necessary). You can also use `↓` to cause `simp` to apply lemmas before simplifying a subterm.

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -173,6 +194,365 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+
+/-- (Noncomputably) bundle the data of a reflexive pair as a functor out of the walking reflexive
+pair -/
+noncomputable def ofIsReflexivePair (f g : A ⟶ B) [IsReflexivePair f g] :
```
It seems this is a particular case of `reflexivePair`, so that it is not necessary to reproduce the construction. You may apply `reflexivePair` to `f`, `g` and `commonSection f g`.

joelriou reviewed (changes requested):
As this is available in mathlib, I would advocate for a definition of the category `WalkingReflexivePair` by using generators and relations as follows:

```lean
import Mathlib.CategoryTheory.PathCategory

namespace CategoryTheory.Limits

inductive WalkingReflexivePairQuiver : Type where
  | zero
  | one

namespace WalkingReflexivePairQuiver

inductive Hom : WalkingReflexivePairQuiver → WalkingReflexivePairQuiver → Type
  | left : Hom zero one
  | right : Hom zero one
  | section_ : Hom one zero

instance : Quiver WalkingReflexivePairQuiver where
  Hom := Hom

abbrev left : zero ⟶ one := Hom.left
abbrev right : zero ⟶ one := Hom.right
abbrev section_ : one ⟶ zero := Hom.section_

inductive homRel : HomRel (Paths WalkingReflexivePairQuiver)
  | section_left : homRel (section_.toPath ≫ left.toPath) (𝟙 _)
  | section_right : homRel (section_.toPath ≫ right.toPath) (𝟙 _)

end WalkingReflexivePairQuiver

def WalkingReflexivePair := Quotient WalkingReflexivePairQuiver.homRel
  deriving Category

end CategoryTheory.Limits
```

It should make a few constructions slightly easier (removing case by case studies). Only the proof that the inclusion from `WalkingParallelPair` is final would need some changes.

kmill commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -173,6 +194,365 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+    | left => f
+    | right => g
+    | reflexion => commonSection f g
+    | right_reflexion => g ≫  (commonSection f g)
+    | left_reflexion => f ≫  (commonSection f g)
```
```suggestion
    | right_reflexion => g ≫ commonSection f g
    | left_reflexion => f ≫ commonSection f g
```

jcommelin commented:
@robin-carlier This PR looks mostly fine, but there are still some places where it needs some work. I see that you have attempted to process some of the review comments. Are you planning to continue working on this PR? Would you like some help from others, or do you want to hand over completely?

robin-carlier commented:
> @robin-carlier This PR looks mostly fine, but there are still some places where it needs some work. I see that you have attempted to process some of the review comments. Are you planning to continue working on this PR? Would you like some help from others, or do you want to hand over completely?

I am taking a break from math (and consequently lean)-related stuff for the time being, so I don’t plan to start working back on this in the near future and I’d gladly hand it over completely if someone wants to work on this or if this content is needed.

robin-carlier commented:
I ended up changing the name of `forgetReflexion` to `inclusionWalkingReflexivePair` in the `WalkingParallelPair` namespace. It leads to rather long names but I’m out of ideas of a better name.

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -3,27 +3,50 @@ Copyright (c) 2020 Bhavik Mehta. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Bhavik Mehta
```
```suggestion
Authors: Bhavik Mehta, Robin Carlier
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,372 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+  | _, _, _, left_reflexion, right_reflexion => left_reflexion
+  | _, _, _, left_reflexion, left_reflexion => left_reflexion
+
+theorem WalkingReflexivePairHom.id_comp
+    {X Y : WalkingReflexivePair} (g : WalkingReflexivePairHom X Y) : comp (id X) g = g := by
+  cases g <;> rfl
+
+theorem WalkingReflexivePairHom.comp_id
+    {X Y : WalkingReflexivePair} (f : WalkingReflexivePairHom X Y) : comp f (id Y) = f := by
+  cases f <;> rfl
+
+theorem WalkingReflexivePairHom.assoc {X Y Z W : WalkingReflexivePair}
+    (f : WalkingReflexivePairHom X Y) (g: WalkingReflexivePairHom Y Z)
+    (h : WalkingReflexivePairHom Z W) : comp (comp f g) h = comp f (comp g h) := by
+  cases f <;> cases g <;> cases h <;> rfl
+
+instance walkingReflexivePairCategory : SmallCategory WalkingReflexivePair where
+  Hom := WalkingReflexivePairHom
+  id := id
+  comp := comp
+  comp_id := comp_id
+  id_comp := id_comp
+  assoc := assoc
```
As these lemmas should not be needed outside of the definition of the category structure, it would suggest doing this:
```suggestion
instance walkingReflexivePairCategory : SmallCategory WalkingReflexivePair where
  Hom := WalkingReflexivePairHom
  id := id
  comp := comp
  comp_id f := by cases f <;> rfl
  id_comp f := by cases f <;> rfl
  assoc f g h := by cases f <;> cases g <;> cases h <;> rfl
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,372 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+open WalkingReflexivePair
+
+/-- The type of morphisms for the diagram indexing reflexive (co)equalizers -/
+inductive WalkingReflexivePairHom : (WalkingReflexivePair → WalkingReflexivePair → Type)
```
All the definition related to the category `WalkingReflexivePair` should be in the same namespace, like `WalkingReflexivePair.Hom`, etc.

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,372 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+namespace WalkingParallelPair
+
+/-- The forgetful functor forgetting the common section -/
+def inclusionWalkingReflexivePair : WalkingParallelPair ⥤ WalkingReflexivePair :=
+  { obj := fun x => match x with
```
In Lean 4, the preferred syntax is:
```suggestion
def inclusionWalkingReflexivePair : WalkingParallelPair ⥤ WalkingReflexivePair where
  obj := fun x => match x with
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,372 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+      | zero => exact e₁
+  cases x with
+    | zero => apply IsConnected.of_induct (j₀ := e₁)
+              intro p h₁ h₂ t
+              rcases t with ⟨l, y, f⟩
```
`intro/rcases` can be combined as:
```suggestion
              rintro p h₁ h₂ ⟨l, y, f⟩
```
(There is another similar situation below.)

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,372 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+instance inclusionWalkingReflexivePair_final : Functor.Final inclusionWalkingReflexivePair := by
+  constructor
+  set e₀ : (StructuredArrow WalkingReflexivePair.one inclusionWalkingReflexivePair) :=
+    StructuredArrow.mk (Y := WalkingParallelPair.zero) (id WalkingReflexivePair.one)
```
```suggestion
    StructuredArrow.mk (Y := WalkingParallelPair.zero) (𝟙 _)
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,372 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+/-- Bundle the data of a parallel pair along with a common section as a functor out of the walking
+reflexive pair -/
+def reflexivePair (f g : A ⟶ B) (s : B ⟶ A) (sl : s ≫ f = 𝟙 B) (sr : s ≫ g = 𝟙 B) :
+    (WalkingReflexivePair ⥤ C) where
```
```suggestion
    WalkingReflexivePair ⥤ C where
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,372 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+theorem walkingReflexivePairHom_id (X : WalkingReflexivePair) :
+    WalkingReflexivePairHom.id X = 𝟙 X := by rfl
+
+@[simp]
+theorem walkingReflexivePairHom_left_comp_reflexion : left.comp reflexion = left_reflexion := rfl
```
As much as possible, one the `Category` instance has been defined, we should use `≫` (`CategoryStruct.comp`) instead of the `comp` defined above. It would take the following form:
```lean
theorem walkingReflexivePairHom_left_comp_reflexion :
  (left ≫ reflexion : one ⟶ one) = left_reflexion := rfl```
```
(The annoying thing is that Lean cannot guess that `left` and `reflexion` are morphisms in a category, so that we need to help them a little bit.)
Actually, I am not convinced that this would be the best normal form, I would rather suggest keeping `left ≫ reflexion` as the normal form (and then `left_reflexion` is part of the internals of the implementation; by the way, `leftCompReflexion` may be a name that fits better our naming conventions!?)

joelriou replied:
However, the following lemmas would be useful:
```lean
@[reassoc (attr := simp)]
theorem reflexion_comp_left : reflexion ≫ left = 𝟙 zero := rfl

@[reassoc (attr := simp)]
theorem reflexion_comp_right : reflexion ≫ right = 𝟙 zero := rfl
```
(With this combination of `reassoc` of `simp`, we should get reasonable normal forms for morphisms in that category.)

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,372 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+
+@[simp]
+theorem reflexivePair_reflexion_comp_left (F : WalkingReflexivePair ⥤ C) :
+    F.map reflexion ≫ F.map left = 𝟙 F.obj zero :=
+  (F.map_id zero).symm.trans (F.map_comp reflexion left)|>.symm
```
Using the `reflexion_comp_left` lemma suggested above: 
```suggestion
        F.map reflexion ≫ F.map left = 𝟙 (F.obj zero) := by
  rw [← F.map_comp, reflexion_comp_left, F.map_id]
```
(Note also the parentheses change.)

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,372 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+
+variable {C : Type u} [Category.{v} C]
+
+@[simp]
```
Doing the following, `F.map reflexion ≫ F.map left ≫f` will simplify into `f`: 
```suggestion
@[reassoc (attr:= simp)]
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,372 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+
+/-- Bundle the data of a parallel pair along with a common section as a functor out of the walking
+reflexive pair -/
+def reflexivePair (f g : A ⟶ B) (s : B ⟶ A) (sl : s ≫ f = 𝟙 B) (sr : s ≫ g = 𝟙 B) :
```
With this, Lean will try to fill in the proofs automatically: 
```suggestion
def reflexivePair (f g : A ⟶ B) (s : B ⟶ A) (sl : s ≫ f = 𝟙 B := by aesop_cat) (sr : s ≫ g = 𝟙 B := by aesop_cat) :
```
Then, you should also add lemmas like
```lean
@[simp] lemma reflexivePair_obj_zero : (reflexivePair f g s sl sr).obj zero = B := rfl
```
and similarly `obj_one`, `map_left`, `map_right` and `map_reflexion`.

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,372 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+
+/-- Any functor out of `WalkingReflexivePair` is isomorphic to the reflexive pair built out of the
+images of its morphisms -/
+def diagramIsoReflexivePair (F : WalkingReflexivePair ⥤ C) :
+    F ≅ reflexivePair (F.map left) (F.map right) (F.map reflexion)
+    ((F.map_id zero).symm.trans (F.map_comp reflexion left)).symm
+    ((F.map_id zero).symm.trans (F.map_comp reflexion right)).symm :=
```
Taking into account suggestions above, I would proceed like this:
```lean
def diagramIsoReflexivePair (F : WalkingReflexivePair ⥤ C) :
    F ≅ reflexivePair (F.map left) (F.map right) (F.map reflexion) :=
  NatIso.ofComponents (fun x ↦ match x with
    | zero => Iso.refl _
    | one => Iso.refl _)
  (fun f ↦ by cases f <;> simp)
```
(As much as possible, I feel it is slightly better not to use `eqToIso`.)
More generally, as you have defined a constructor for functors from the category `WalkingReflexivePair`, it would be nice to introduce also a constructor for natural transformations (and then natural isomorphisms) between objets in `WalkingReflexivePair ⥤ C`. Then, `diagramIsoReflexivePair` would appear as a special case of this.

joelriou commented:
This PR needs some modifications so as to fit more into the mathlib conventions, but the results look very nice to me.

robin-carlier commented:
Thank you for the extensive review and the numerous tricks I did not know about!
I have implemented your suggestions, and added a few new constructors to bundles functors from or towards a term of the form `reflexivePair …`.
To avoid cluttering things too much, I decided not to put a constructor for natural transforms between two `reflexivePair`, because it can already be constructed with `mkNatTrans` and `mkNatTrans'` if needed.

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,492 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+open WalkingReflexivePair
+
+namespace WalkingReflexivePair
+/-- The type of morphisms for the diagram indexing reflexive (co)equalizers -/
```
```suggestion

/-- The type of morphisms for the diagram indexing reflexive (co)equalizers -/
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,492 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+  comp_id := by intro _ _ f; cases f <;> rfl
+  id_comp := by intro _ _ f; cases f <;> rfl
+  assoc := by intro _ _ _ _ f g h; cases f <;> cases g <;> cases h <;> rfl
+
```
```suggestion
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,492 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+open Hom
+
+@[simp]
+lemma Hom_id (X : WalkingReflexivePair) :
```
```suggestion
lemma Hom.id_eq (X : WalkingReflexivePair) :
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,492 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+
+@[reassoc (attr := simp)]
+lemma reflexion_comp_right : reflexion ≫ right = 𝟙 zero := rfl
+
```
```suggestion

```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,491 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+  have h : Inhabited (StructuredArrow x inclusionWalkingReflexivePair) := by
+    · constructor
+      cases x with
+      | one => exact e₀
+      | zero => exact e₁
+  cases x with
+    | zero => apply IsConnected.of_induct (j₀ := e₁)
+              rintro p h₁ h₂ ⟨_, y, f⟩
+              cases y <;> cases f
+              · set r : StructuredArrow .zero inclusionWalkingReflexivePair :=
+                  StructuredArrow.mk (Y := zero) reflexion
+                change r ∈ p
+                suffices f : r ⟶  e₁ by exact (h₂ f).mpr h₁
+                exact StructuredArrow.homMk .left
+              · exact h₁
+    | one => apply IsConnected.of_induct (j₀ := e₀)
+             rintro p h₁ h₂ ⟨_, y, f⟩
+             set rₗ : StructuredArrow WalkingReflexivePair.one inclusionWalkingReflexivePair :=
+                StructuredArrow.mk (Y := one) left
+             set rᵣ : StructuredArrow WalkingReflexivePair.one inclusionWalkingReflexivePair :=
+                StructuredArrow.mk (Y := one) right
+             have hrₗ : rₗ ∈ p := by
+              · suffices f : e₀ ⟶  rₗ by exact (h₂ f).mp h₁
+                exact StructuredArrow.homMk .left
+             have hrᵣ : rᵣ ∈ p := by
+              · suffices f : e₀ ⟶  rᵣ by exact (h₂ f).mp h₁
+                exact StructuredArrow.homMk .right
+             cases y <;> cases f
+             rotate_right 3
+             · exact h₁
+             · exact hrₗ
+             · exact hrᵣ
+             · set v : StructuredArrow .one inclusionWalkingReflexivePair :=
+                StructuredArrow.mk (Y := zero) leftCompReflexion
+               suffices f : v ⟶  rₗ by exact (h₂ f).mpr hrₗ
+               exact StructuredArrow.homMk .left
+             · set v : StructuredArrow .one inclusionWalkingReflexivePair :=
+                StructuredArrow.mk (Y := .zero) rightCompReflexion
+               suffices f : v ⟶  rᵣ by exact (h₂ f).mpr hrᵣ
+               exact StructuredArrow.homMk .right
```
The syntax can be improved:
```suggestion
instance (X : WalkingReflexivePair) :
    Nonempty (StructuredArrow X inclusionWalkingReflexivePair) := by
  cases X with
  | zero => exact ⟨StructuredArrow.mk (Y := one) (𝟙 _)⟩
  | one => exact ⟨StructuredArrow.mk (Y := zero) (𝟙 _)⟩

open WalkingReflexivePair.Hom in
instance (X : WalkingReflexivePair) :
    IsConnected (StructuredArrow X inclusionWalkingReflexivePair) := by
  cases X with
  | zero =>
      refine IsConnected.of_induct  (j₀ := StructuredArrow.mk (Y := one) (𝟙 _)) ?_
      rintro p h₁ h₂ ⟨⟨⟨⟩⟩, (_ | _), ⟨_⟩⟩
      · exact (h₂ (StructuredArrow.homMk .left)).2 h₁
      · exact h₁
  | one =>
      refine IsConnected.of_induct  (j₀ := StructuredArrow.mk (Y := zero) (𝟙 _))
        (fun p h₁ h₂ ↦ ?_)
      have hₗ : StructuredArrow.mk left ∈ p := (h₂ (StructuredArrow.homMk .left)).1 h₁
      have hᵣ : StructuredArrow.mk right ∈ p := (h₂ (StructuredArrow.homMk .right)).1 h₁
      rintro ⟨⟨⟨⟩⟩, (_ | _), ⟨_⟩⟩
      · exact (h₂ (StructuredArrow.homMk .left)).2 hₗ
      · exact (h₂ (StructuredArrow.homMk .right)).2 hᵣ
      all_goals assumption

/-- The inclusion functor is a final functor -/
instance inclusionWalkingReflexivePair_final : Functor.Final inclusionWalkingReflexivePair where
  out := inferInstance
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,491 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+variable {A B : C}
+variable (f g : A ⟶ B) (s : B ⟶ A) (sl : s ≫ f = 𝟙 B) (sr : s ≫ g = 𝟙 B)
+
+@[simp] lemma obj_zero : (reflexivePair f g s sl sr).obj zero = B := rfl
```
These lemmas should preferably be stated just after the definition of `reflexivePair`. The standard names should be `reflexivePair_obj_zero`, etc.

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,491 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+
+section NatTrans
+
+variable (F : WalkingReflexivePair ⥤ C)
```
```suggestion
variable {F : WalkingReflexivePair ⥤ C}
```
This seems better to me for the next three definition. `varaiable (F)` should then be added after `mkNatIso`.

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,491 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+                                               reassoc_of% h₂]
+
+/-- Constructor for natural isomorphisms with a `reflexivePair`. -/
+@[simps!]
+def mkNatIso (u : F.obj one ≅ A) (v : F.obj zero ≅ B)
+    (h₁ : (F.map left) ≫ v.hom = u.hom ≫ f := by aesop_cat)
+    (h₂ : (F.map right) ≫ v.hom = u.hom ≫ g := by aesop_cat)
+    (h₃ : (F.map reflexion) ≫ u.hom = v.hom ≫ s := by aesop_cat) :
+    F ≅ (reflexivePair f g s sl sr) :=
+  NatIso.ofComponents (fun x ↦ match x with
+    | zero => v
+    | one => u)
+  (by intro _ _ f
+      cases f with
+        | id => simp
+        | left => simp [h₁]
+        | right => simp [h₂]
+        | reflexion => simp [h₃]
+        | leftCompReflexion => dsimp
+                               simp only [Functor.map_comp, obj_one, obj_zero,
+                                map_right, map_reflexion, Category.assoc, h₃,
+                                reassoc_of% h₁]
+        | rightCompReflexion => dsimp
+                                simp only [Functor.map_comp, obj_one, obj_zero,
+                                 map_left, map_reflexion, Category.assoc, h₃,
+                                 reassoc_of% h₂])
```
Arguably, this would be slightly better:
```suggestion
@[simps!]
/-- Constructor for natural isomorphisms with a `reflexivePair`. -/
@[simps]
def mkNatIso (u : F.obj one ≅ A) (v : F.obj zero ≅ B)
    (h₁ : (F.map left) ≫ v.hom = u.hom ≫ f := by aesop_cat)
    (h₂ : (F.map right) ≫ v.hom = u.hom ≫ g := by aesop_cat)
    (h₃ : (F.map reflexion) ≫ u.hom = v.hom ≫ s := by aesop_cat) :
    F ≅ reflexivePair f g s sl sr where
  hom := mkNatTrans u.hom v.hom 
  inv := mkNatTrans' u.inv v.inv
    (by rw [← cancel_epi u.hom, u.hom_inv_id_assoc, ← reassoc_of% h₁, v.hom_inv_id,
            Category.comp_id])
    (by rw [← cancel_epi u.hom, u.hom_inv_id_assoc, ← reassoc_of% h₂, v.hom_inv_id,
            Category.comp_id])
    (by rw [← cancel_epi v.hom, v.hom_inv_id_assoc, ← reassoc_of% h₃, u.hom_inv_id,
            Category.comp_id])
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,440 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+section NatTrans
+
+/-- A constructor for natural transforms to a diagram of the form `reflexivePair f g s`. -/
+def mkNatTrans {F G : WalkingReflexivePair ⥤ C}
```
`variable {F G : WalkingReflexivePair ⥤ C}` should be inserted before these declarations.

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,440 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+
+/-- Constructor for natural isomorphisms with a `reflexivePair`. -/
+@[simps!]
+def mkNatIso (u : F.obj one ≅ G.obj one) (v : F.obj zero ≅ G.obj zero)
```
It would be better to use the same order of parameters as for `mkNatTrans`. (The proof could be exactly the same, or by constructing `hom`, and `inv` using `mkNatTrans` as I suggested.)
```suggestion
def mkNatIso (e₀ : F.obj zero ≅ G.obj zero) (e₁ : F.obj one ≅ G.obj one)
```

joelriou replied:
(Same for `whiskerRightMkNatTrans`.)

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,440 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+    (u : F.obj zero ⟶ G.obj zero) (v : F.obj one ⟶ G.obj one)
+    {h₁ : F.map left ≫ u = v ≫ G.map left}
+    {h₂ : F.map right ≫ u = v ≫ G.map right}
+    {h₃ : F.map reflexion ≫ v = u ≫ G.map reflexion}
+    (H : C ⥤ D) :
+      (whiskerRight (mkNatTrans u v : F ⟶ G) H) = mkNatTrans
+            (H.map u) (H.map v)
+            (by simp only [Functor.comp_obj, Functor.comp_map, ← Functor.map_comp, h₁])
+            (by simp only [Functor.comp_obj, Functor.comp_map, ← Functor.map_comp, h₂])
+            (by simp only [Functor.comp_obj, Functor.comp_map, ← Functor.map_comp, h₃]) := by
```
```suggestion
    {h₃ : F.map reflexion ≫ v = u ≫ G.map reflexion} (H : C ⥤ D) :
    (whiskerRight (mkNatTrans u v : F ⟶ G) H) =
      mkNatTrans (H.map u) (H.map v)
          (by simp only [Functor.comp_obj, Functor.comp_map, ← Functor.map_comp, h₁])
          (by simp only [Functor.comp_obj, Functor.comp_map, ← Functor.map_comp, h₂])
          (by simp only [Functor.comp_obj, Functor.comp_map, ← Functor.map_comp, h₃]) := by
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,440 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+variable {F : WalkingReflexivePair ⥤ C}
+
+/-- The tail morphism of a reflexive cofork. -/
+abbrev π (G : ReflexiveCofork F) := G.ι.app zero
```
For readibility, it is advisable to announce the type of defs/abbrevs:
```suggestion
abbrev π (G : ReflexiveCofork F) : F.obj zero ⟶ G.pt := G.ι.app zero
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,440 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+
+@[simp]
+lemma app_one_eq_π (G : ReflexiveCofork F) : G.ι.app zero = G.π := rfl
+
```
It would be nice to add a constructor for `ReflexiveCofork`:
```lean
@[simps pt]
def mk {X : C} (π : F.obj zero ⟶ X) (h : F.map left ≫ π = F.map right ≫ π) :
    ReflexiveCofork F where
  pt := X
  ι := reflexivePair.mkNatTrans π (F.map left ≫ π)

@[simp]
lemma mk_π {X : C} (π : F.obj zero ⟶ X) (h : F.map left ≫ π = F.map right ≫ π) :
    (mk π h).π = π := rfl
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,438 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+    {h₂ : F.map right ≫ e₀ = e₁ ≫ G.map right}
+    {h₃ : F.map reflexion ≫ e₁ = e₀ ≫ G.map reflexion}
+    {D : Type u₂} [Category.{v₂} D] (H : C ⥤ D) :
+      (whiskerRight (mkNatTrans e₀ e₁ : F ⟶ G) H) =
+        mkNatTrans (H.map e₀) (H.map e₁)
+            (by simp only [Functor.comp_obj, Functor.comp_map, ← Functor.map_comp, h₁])
+            (by simp only [Functor.comp_obj, Functor.comp_map, ← Functor.map_comp, h₂])
+            (by simp only [Functor.comp_obj, Functor.comp_map, ← Functor.map_comp, h₃]) := by
```
```suggestion
    whiskerRight (mkNatTrans e₀ e₁ : F ⟶ G) H =
      mkNatTrans (H.map e₀) (H.map e₁)
          (by simp only [Functor.comp_obj, Functor.comp_map, ← Functor.map_comp, h₁])
          (by simp only [Functor.comp_obj, Functor.comp_map, ← Functor.map_comp, h₂])
          (by simp only [Functor.comp_obj, Functor.comp_map, ← Functor.map_comp, h₃]) := by
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,438 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+lemma app_one_eq_π (G : ReflexiveCofork F) : G.ι.app zero = G.π := rfl
+
+/-- The underlying `Cofork` of a `ReflexiveCofork`. -/
+def toCofork (G : ReflexiveCofork F) : (Cofork (F.map left) (F.map right)) :=
```
```suggestion
def toCofork (G : ReflexiveCofork F) : Cofork (F.map left) (F.map right) :=
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,438 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+
+/-- The underlying `Cofork` of a `ReflexiveCofork`. -/
+def toCofork (G : ReflexiveCofork F) : (Cofork (F.map left) (F.map right)) :=
+  Cocones.precompose (diagramIsoParallelPair (_ ⋙ F)).symm.hom |>.obj
+    <| G.whisker WalkingParallelPair.inclusionWalkingReflexivePair
```
Linebreaks should be after infix operators:
```suggestion
  Cocones.precompose (diagramIsoParallelPair (_ ⋙ F)).symm.hom |>.obj <|
    G.whisker WalkingParallelPair.inclusionWalkingReflexivePair
```
(Same below.)

robin-carlier replied:
Got it, I thought the `<|` and `|>` were following the rules for parentheses rather than operators. I’ll try to be careful next time.

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,438 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+section NatTrans
+
+variable {F G: WalkingReflexivePair ⥤ C}
+/-- A constructor for natural transforms to a diagram of the form `reflexivePair f g s`. -/
```
/-- A constructor for natural transformations between functors from `WalkingReflexivePair`. -/

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,438 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+        Category.assoc]
+
+@[simp]
+lemma mkNatTrans_app_zero (e₀ : F.obj zero ⟶ G.obj zero) (e₁ : F.obj one ⟶ G.obj one)
```
The variables for this lemma and the next are the same as above, so that you should put these three definitions/lemmas in a `section` and use `variable ...`.

robin-carlier replied:
Is there a syntax to make the autoparam `:= by aesop_cat` apply to only the definition `mkNatTrans` while putting these variables in a `variable ...`? Or should I just leave the autoparam even for the lemmas?

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,438 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+      (by simp only [← Functor.map_comp, sr, Functor.map_id]) :=
+  mkNatIso (Iso.refl _) (Iso.refl _)
+
+lemma whiskerRightMkNatTrans {F G: WalkingReflexivePair ⥤ C}
```
```suggestion
lemma whiskerRightMkNatTrans {F G : WalkingReflexivePair ⥤ C}
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,438 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+section
+section NatTrans
+
+variable {F G: WalkingReflexivePair ⥤ C}
```
```suggestion
variable {F G : WalkingReflexivePair ⥤ C}
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,438 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+lemma app_one_eq_π (G : ReflexiveCofork F) : G.ι.app zero = G.π := rfl
+
+/-- The underlying `Cofork` of a `ReflexiveCofork`. -/
+def toCofork (G : ReflexiveCofork F) : Cofork (F.map left) (F.map right) :=
+  Cocones.precompose (diagramIsoParallelPair (_ ⋙ F)).symm.hom |>.obj <|
+    G.whisker WalkingParallelPair.inclusionWalkingReflexivePair
+
+@[simp]
+lemma toCofork.π (G : ReflexiveCofork F) : Cofork.π (G.toCofork) = G.π := by
+  dsimp only [toCofork, Cofork.π]
+  aesop_cat
```
I would favour a more down to earth definition here:
```suggestion
abbrev toCofork (G : ReflexiveCofork F) : Cofork (F.map left) (F.map right) :=
  Cofork.ofπ G.π (by simp)
```

Then, `inclusionWalkingReflexivePairEquiv_obj` below become an iso (use `Cofork.ext`) instead of an equality, and the definition of `ReflectiveCofork.isColimitEquiv` would have to be changed, using the composition with `IsColimit.equivIsoColimit (inclusionWalkingReflexivePairEquivObjIso G).symm)`.

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,435 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+
+instance ofIsReflexivePairHasColimit_of_hasCoequalizer [HasCoequalizer f g] :
+    HasColimit (ofIsReflexivePair f g) := by
+  suffices _ : HasColimit <|
+    WalkingParallelPair.inclusionWalkingReflexivePair ⋙ (ofIsReflexivePair f g) by
+    · apply Functor.Final.hasColimit_of_comp WalkingParallelPair.inclusionWalkingReflexivePair
+  exact Limits.hasColimitOfIso <| inclusionWalkingReflexivePairOfIsReflexivePairIso f g
```
```suggestion
  have := Limits.hasColimitOfIso <| inclusionWalkingReflexivePairOfIsReflexivePairIso f g
  apply Functor.Final.hasColimit_of_comp WalkingParallelPair.inclusionWalkingReflexivePair
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,435 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+
+instance reflexivePair_hasColimit_of_hasCoequalizer
+    [h : HasCoequalizer (F.map left) (F.map right)] : HasColimit F := by
+  suffices _ : HasColimit (WalkingParallelPair.inclusionWalkingReflexivePair ⋙ F) by
+    · apply Functor.Final.hasColimit_of_comp WalkingParallelPair.inclusionWalkingReflexivePair
+  exact @Limits.hasColimitOfIso _ _ _ _ _ _ h <| diagramIsoParallelPair _
```
```suggestion
  have := Limits.hasColimitOfIso (show _ ≅ parallelPair (F.map left) (F.map right) from
    diagramIsoParallelPair (WalkingParallelPair.inclusionWalkingReflexivePair ⋙ F))
  apply Functor.Final.hasColimit_of_comp WalkingParallelPair.inclusionWalkingReflexivePair
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,435 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+      (diagramIsoParallelPair (WalkingParallelPair.inclusionWalkingReflexivePair ⋙ F))).symm
+
+instance reflexivePair_hasColimit_of_hasCoequalizer
+    [h : HasCoequalizer (F.map left) (F.map right)] : HasColimit F := by
```
```suggestion
    [HasCoequalizer (F.map left) (F.map right)] : HasColimit F := by
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,435 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+
+open WalkingReflexivePair WalkingReflexivePair.Hom
+
+variable {F : WalkingReflexivePair ⥤ C}
```
I would declare `F` as an explicit variable here:
```suggestion
variable (F : WalkingReflexivePair ⥤ C)
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,435 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+    (Cocones.precomposeEquivalence
+      (diagramIsoParallelPair (WalkingParallelPair.inclusionWalkingReflexivePair ⋙ F))).symm
+
+instance reflexivePair_hasColimit_of_hasCoequalizer
```
The following lemma could be useful:
```
lemma hasReflexiveCoequalizer_iff_hasCoequalizer :
    HasColimit F ↔ HasCoequalizer (F.map left) (F.map right) := by
  simpa only [hasColimit_iff_hasInitial_cocone]
    using Equivalence.hasInitial_iff (inclusionWalkingReflexivePairEquivCocone F)
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,435 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+
+/-- The colimit of a functor out of the walking reflexive pair is the same as the colimit of the
+underlying parallel pair. -/
+noncomputable def inclusionWalkingReflexivePairColimitEquiv
+    [HasCoequalizer (F.map left) (F.map right)] :
+    colimit F ≅ coequalizer (F.map left) (F.map right) :=
+  (Functor.Final.colimitIso WalkingParallelPair.inclusionWalkingReflexivePair F).symm.trans <|
+    HasColimit.isoOfNatIso <|
+      diagramIsoParallelPair <|
+        WalkingParallelPair.inclusionWalkingReflexivePair ⋙ F
```
It is probably better to rely on the API defined just above:
```suggestion
noncomputable def reflexiveCoqualizerIsoCoequalizer
    [HasCoequalizer (F.map left) (F.map right)] :
    colimit F ≅ coequalizer (F.map left) (F.map right) :=
  ((ReflexiveCofork.isColimitEquiv _).symm (colimit.isColimit F)).coconePointUniqueUpToIso
    (colimit.isColimit _)
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,435 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+walking reflexive pair. -/
+theorem hasReflexiveCoequalizers_iff :
+    HasColimitsOfShape WalkingReflexivePair C ↔ HasReflexiveCoequalizers C := by
+  constructor
+  · intro _
+    constructor
+    intro _ _ f g h₁
+    set F := @ofIsReflexivePair _ _ _ _ _ _ h₁
+    exact @Limits.hasColimitOfIso _ _ _ _ _ _
+      (Functor.Final.comp_hasColimit _)
+      (diagramIsoParallelPair (WalkingParallelPair.inclusionWalkingReflexivePair ⋙ F)).symm
+  · intro _
+    exact ⟨by infer_instance⟩
```
```suggestion
  ⟨fun _ ↦ ⟨fun _ _ f g _ ↦ (hasReflexiveCoequalizer_iff_hasCoequalizer
      (reflexivePair f g (commonSection f g))).1 inferInstance⟩,
    fun _ ↦ ⟨inferInstance⟩⟩
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,431 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+  (h₂ : F.map right ≫ e₀ = e₁ ≫ G.map right := by aesop_cat)
+  (h₃ : F.map reflexion ≫ e₁ = e₀ ≫ G.map reflexion := by aesop_cat)
+
+/-- A constructor for natural transformations betweens functors from `WalkingReflexivePair`. -/
```
```suggestion
/-- A constructor for natural transformations between functors from `WalkingReflexivePair`. -/
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,431 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+abbrev toCofork (G : ReflexiveCofork F) : Cofork (F.map left) (F.map right) :=
+  Cofork.ofπ G.π (by simp)
+
+@[simp]
+lemma toCofork.π (G : ReflexiveCofork F) : Cofork.π (G.toCofork) = G.π := by
+  dsimp only [toCofork, Cofork.π]
+  aesop_cat
+
```
The name should have been `toCofork_π`, but `Cofork.ofπ` behaves better as an abbrev, which makes the lemma unnecessary:
```suggestion
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,431 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+    (Cocones.precomposeEquivalence
+      (diagramIsoParallelPair (WalkingParallelPair.inclusionWalkingReflexivePair ⋙ F))).symm
+
+@[inherit_doc inclusionWalkingReflexivePairEquivCocone]
+def inclusionWalkingReflexivePairEquiv_obj (G : ReflexiveCofork F) :
```
```suggestion
@[inherit_doc inclusionWalkingReflexivePairEquivCocone]
def inclusionWalkingReflexivePairEquivObjIso (G : ReflexiveCofork F) :
```

joelriou replied:
```suggestion
def inclusionWalkingReflexivePairEquivObjIso (G : ReflexiveCofork F) :
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,431 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+    (Cocones.precomposeEquivalence
+      (diagramIsoParallelPair (WalkingParallelPair.inclusionWalkingReflexivePair ⋙ F))).symm
+
+@[inherit_doc inclusionWalkingReflexivePairEquivCocone]
```
Here, we need a specific docstring saying this is an isomorphism.

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,431 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+  Cofork.ext (Iso.refl _)
+    (by simp [inclusionWalkingReflexivePairEquivCocone, Cofork.π])
+
+instance reflexivePair_hasColimit_of_hasCoequalizer
```
The point of my lemma `hasReflexiveCoequalizer_iff_hasCoequalizer` was to facilitate the proofs of `reflexivePair_hasColimit_of_hasCoequalizer` and `ofIsReflexivePairHasColimit_of_hasCoequalizer`.

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,431 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+    [HasCoequalizer (F.map left) (F.map right)] :
+    colimit F ≅ coequalizer (F.map left) (F.map right) :=
+  ((ReflexiveCofork.isColimitEquiv _ _).symm (colimit.isColimit F)).coconePointUniqueUpToIso
+    (colimit.isColimit _)variable {A B : C} {f g : A ⟶ B} [IsReflexivePair f g]
```
```suggestion
    (colimit.isColimit _)
    
variable {A B : C} {f g : A ⟶ B} [IsReflexivePair f g]
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,431 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+
+/-- The colimit of a functor out of the walking reflexive pair is the same as the colimit of the
+underlying parallel pair. -/
+def inclusionWalkingReflexivePairColimitEquiv
```
The name should contain `Iso`:
```suggestion
def reflexiveCoequalizerIsoCoequalizer
```

joelriou replied:
You should also add two simp lemmas expressing compatibilities of this isomorphism with the canonical maps to the colimit and to the coequalizer. Names would be `ι_reflexiveCoequalizerIsoCoequalizer_hom` and `π_reflexiveCoequalizerIsoCoequalizer_inv`. (Same remarls for `ofIsReflexivePairColimitEquiv` below.)

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,431 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+  ((ReflexiveCofork.isColimitEquiv _ _).symm (colimit.isColimit F)).coconePointUniqueUpToIso
+    (colimit.isColimit _)variable {A B : C} {f g : A ⟶ B} [IsReflexivePair f g]
+
+instance ofIsReflexivePairHasColimit_of_hasCoequalizer [HasCoequalizer f g] :
```
`HasColimit` is a `Prop`, then this instance is a "lemma" (not a "def"), so that according to our naming conventions we should use `_` to connect the different elements in the name:
```suggestion
instance ofIsReflexivePair_hasColimit_of_hasCoequalizer [HasCoequalizer f g] :
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,431 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+
+/-- The coequalizer of a reflexive pair can be promoted to the colimit of a diagram out of the
+walking reflexive pair -/
+def ofIsReflexivePairColimitEquiv [HasCoequalizer f g] :
+    colimit (ofIsReflexivePair f g) ≅ coequalizer f g :=
+  (Functor.Final.colimitIso _ _).symm.trans <|
+    HasColimit.isoOfNatIso <| inclusionWalkingReflexivePairOfIsReflexivePairIso _ _
```
The name should be changed. For the "proof", it would be better to apply `reflexiveCoequalizerIsoCoequalizer` to `ofIsReflexivePair f g` instead of proving it again.

robin-carlier commented:
I renamed `inclusionWalkingReflexivePairEquivCocone` to `reflexiveCoforkEquivCofork`, and `ofIsReflexivePairColimitEquiv` to `reflexiveCoequalizerIsoCoequalizer`. Since we’are adding simps lemmas, I took the liberty to also add `ofIsReflexivePair_map_left`, `ofIsReflexivePair_map_right`, `reflexiveCoforkEquivCofork_functor_obj_π` and `reflexiveCoforkEquivCofork_inverse_obj_π` for completeness. I couldn’t golf more the proof of the latter unfortunately.

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,486 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+    (h₂ : F.map right ≫ e₀.hom = e₁.hom ≫ G.map right := by aesop_cat)
+    (h₃ : F.map reflexion ≫ e₁.hom = e₀.hom ≫ G.map reflexion := by aesop_cat) :
+    F ≅ G where
+      hom := mkNatTrans e₀.hom e₁.hom
+      inv := mkNatTrans e₀.inv e₁.inv
+        (by rw [← cancel_epi e₁.hom, e₁.hom_inv_id_assoc, ← reassoc_of% h₁, e₀.hom_inv_id,
+            Category.comp_id])
+        (by rw [← cancel_epi e₁.hom, e₁.hom_inv_id_assoc, ← reassoc_of% h₂, e₀.hom_inv_id,
+            Category.comp_id])
+        (by rw [← cancel_epi e₀.hom, e₀.hom_inv_id_assoc, ← reassoc_of% h₃, e₁.hom_inv_id,
+            Category.comp_id])
+      hom_inv_id := by ext x; cases x <;> simp
+      inv_hom_id := by ext x; cases x <;> simp
```
```suggestion
  hom := mkNatTrans e₀.hom e₁.hom
  inv := mkNatTrans e₀.inv e₁.inv
        (by rw [← cancel_epi e₁.hom, e₁.hom_inv_id_assoc, ← reassoc_of% h₁, e₀.hom_inv_id,
            Category.comp_id])
        (by rw [← cancel_epi e₁.hom, e₁.hom_inv_id_assoc, ← reassoc_of% h₂, e₀.hom_inv_id,
            Category.comp_id])
        (by rw [← cancel_epi e₀.hom, e₀.hom_inv_id_assoc, ← reassoc_of% h₃, e₁.hom_inv_id,
            Category.comp_id])
  hom_inv_id := by ext x; cases x <;> simp
  inv_hom_id := by ext x; cases x <;> simp
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,486 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+  ((ReflexiveCofork.isColimitEquiv _ _).symm (colimit.isColimit F)).coconePointUniqueUpToIso
+    (colimit.isColimit _)
+
+@[simp]
```
```suggestion
@[reassoc (attr := simp)]
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,486 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+  IsColimit.comp_coconePointUniqueUpToIso_hom
+    ((ReflexiveCofork.isColimitEquiv F _).symm _) _ WalkingParallelPair.one
+
+@[simp]
```
```suggestion
@[reassoc (attr := simp)]
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,486 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+  @reflexiveCoequalizerIsoCoequalizer _ _ (ofIsReflexivePair f g) h
+
+
+@[simp]
```
```suggestion
@[reassoc (attr := simp)]
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,486 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+    colimit.ι (ofIsReflexivePair f g) zero ≫ colimitOfIsReflexivePairIsoCoequalizer.hom =
+      coequalizer.π f g := @ι_reflexiveCoequalizerIsoCoequalizer_hom _ _ _ h
+
+@[simp]
```
```suggestion
@[reassoc (attr := simp)]
```

joelriou commented on Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean:
```diff
@@ -148,6 +171,486 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+  aesop_cat
+
+@[simp]
+lemma inclusionWalkingReflexivePairEquivCocone_inverse_obj_π
```
```suggestion
lemma reflexiveCoforkEquivCofork_inverse_obj_π
```
I will try to see if there is a better proof...

joelriou replied:
Could you add the following lemma in `Functor.Final`:
```lean
/-- Alternative equational lemma for `(extendCocone c).ι.app` in case a lift of the object
is given explicitly. -/
lemma extendCocone_obj_ι_app' (c : Cocone (F ⋙ G)) {X : D} {Y : C} (f : X ⟶ F.obj Y) :
    (extendCocone.obj c).ι.app X = G.map f ≫ c.ι.app Y := by
  sorry
```
and then the proof here can be:
```lean
@[simp]
lemma reflexiveCoforkEquivCofork_inverse_obj_π
    (G : Cofork (F.map left) (F.map right)) :
    ((reflexiveCoforkEquivCofork F).inverse.obj G).π = G.π := by
  dsimp only [reflexiveCoforkEquivCofork, Equivalence.symm, Equivalence.trans,
    ReflexiveCofork.π, Cocones.precomposeEquivalence, Cocones.precompose,
    Functor.comp, Functor.Final.coconesEquiv]
  rw [Functor.Final.extendCocone_obj_ι_app' (Y := .one) (f := 𝟙 zero)]
  simp
```

robin-carlier commented:
This has been rebased onto https://github.com/leanprover-community/mathlib4/pull/17723, on which it now depends on.

joelriou commented:
I added a space just before merging #17723, so that you need to `git merge origin` and keep the incoming version.

joelriou commented on Mathlib/SetTheory/Ordinal/Arithmetic.lean:
```diff
@@ -758,6 +758,9 @@ theorem isNormal_mul_right {a : Ordinal} (h : 0 < a) : IsNormal (a * ·) :=
 @[deprecated isNormal_mul_right (since := "2024-10-11")]
 alias mul_isNormal := isNormal_mul_right
 
+@[deprecated isNormal_mul_right (since := "2024-10-11")]
+alias mul_isNormal := isNormal_mul_right
+
```
```suggestion
```

joelriou commented on Mathlib/SetTheory/Ordinal/Exponential.lean:
```diff
@@ -108,6 +108,9 @@ theorem isNormal_opow {a : Ordinal} (h : 1 < a) : IsNormal (a ^ ·) :=
 @[deprecated isNormal_opow (since := "2024-10-11")]
 alias opow_isNormal := isNormal_opow
 
+@[deprecated isNormal_opow (since := "2024-10-11")]
+alias opow_isNormal := isNormal_opow
+
```
```suggestion
```
